File: client/python/translate.py
"""Pure MagicCDP <-> CDP translation helpers for the Python client."""

import json

BINDING_PREFIX = "__MagicCDP_"

DEFAULT_CLIENT_ROUTES = {
    "Magic.*": "service_worker",
    "Custom.*": "service_worker",
    "*.*": "direct_cdp",
}


def binding_name_for(event_name: str) -> str:
    return BINDING_PREFIX + event_name.replace(".", "_")


def event_name_for(binding_name: str):
    if not binding_name.startswith(BINDING_PREFIX):
        return None
    return binding_name[len(BINDING_PREFIX):].replace("_", ".")


def route_for(method: str, routes: dict) -> str:
    routes = routes or {}
    if method in routes:
        return routes[method]
    best_prefix_len = -1
    best_route = None
    for pattern, route in routes.items():
        if pattern == "*.*" or not pattern.endswith(".*"):
            continue
        prefix = pattern[:-1]
        if method.startswith(prefix) and len(prefix) > best_prefix_len:
            best_prefix_len = len(prefix)
            best_route = route
    if best_route is not None:
        return best_route
    if "*.*" in routes:
        return routes["*.*"]
    return "direct_cdp"


def _eval_params(expression: str):
    return {
        "expression": expression,
        "awaitPromise": True,
        "returnByValue": True,
        "allowUnsafeEvalBlockedByCSP": True,
    }


def _wrap_magic_evaluate(params: dict, session_id: str):
    expression = params["expression"]
    user_params = params.get("params", {})
    cdp_session_id = params.get("cdpSessionId") or session_id
    return _eval_params(
        "(async () => {\n"
        f"  const params = {json.dumps(user_params)};\n"
        f"  const cdp = globalThis.MagicCDP.attachToSession({json.dumps(cdp_session_id)});\n"
        "  const MagicCDP = globalThis.MagicCDP;\n"
        "  const chrome = globalThis.chrome;\n"
        f"  const value = ({expression});\n"
        "  return typeof value === 'function' ? await value(params) : value;\n"
        "})()"
    )


def _wrap_magic_add_custom_command(params: dict):
    return _eval_params(
        "(() => {\n"
        "  return globalThis.MagicCDP.addCustomCommand({\n"
        f"    name: {json.dumps(params['name'])},\n"
        f"    paramsSchema: {json.dumps(params.get('paramsSchema'))},\n"
        f"    resultSchema: {json.dumps(params.get('resultSchema'))},\n"
        f"    expression: {json.dumps(params['expression'])},\n"
        "    handler: async (params, meta) => {\n"
        "      const cdp = globalThis.MagicCDP.attachToSession(meta.cdpSessionId);\n"
        "      const MagicCDP = globalThis.MagicCDP;\n"
        "      const chrome = globalThis.chrome;\n"
        f"      const handler = ({params['expression']});\n"
        "      return await handler(params || {});\n"
        "    },\n"
        "  });\n"
        "})()"
    )


def _wrap_magic_add_custom_event(params: dict):
    return _eval_params(
        "globalThis.MagicCDP.addCustomEvent({\n"
        f"  name: {json.dumps(params['name'])},\n"
        f"  bindingName: {json.dumps(binding_name_for(params['name']))},\n"
        f"  payloadSchema: {json.dumps(params.get('payloadSchema'))},\n"
        "})"
    )


def _wrap_custom_command(method: str, params: dict, session_id: str):
    return _eval_params(
        f"globalThis.MagicCDP.handleCommand("
        f"{json.dumps(method)}, {json.dumps(params)}, "
        f"{json.dumps({'cdpSessionId': session_id})})"
    )


def _wrap_magic_configure(config: dict):
    return _eval_params(
        "(async () => {\n"
        "  const deadline = Date.now() + 5000;\n"
        "  while (!globalThis.MagicCDP?.configure && Date.now() < deadline) {\n"
        "    await new Promise(resolve => setTimeout(resolve, 25));\n"
        "  }\n"
        "  if (!globalThis.MagicCDP?.configure) throw new Error('MagicCDP service worker global is unavailable.');\n"
        f"  return globalThis.MagicCDP.configure({json.dumps(config)});\n"
        "})()"
    )


def _translate_service_worker_command(method: str, params: dict, session_id: str):
    if method == "Magic.addCustomEvent":
        return [
            {"method": "Runtime.addBinding", "params": {"name": binding_name_for(params["name"])}},
            {
                "method": "Runtime.evaluate",
                "params": _wrap_magic_add_custom_event(params),
                "unwrap": "evaluate",
            },
        ]
    if method == "Magic.evaluate":
        runtime_params = _wrap_magic_evaluate(params, session_id)
    elif method == "Magic.addCustomCommand":
        runtime_params = _wrap_magic_add_custom_command(params)
    else:
        runtime_params = _wrap_custom_command(method, params, params.get("cdpSessionId") or session_id)
    return [{"method": "Runtime.evaluate", "params": runtime_params, "unwrap": "evaluate"}]


def translate_client_command(method: str, params=None, *, routes=None, cdp_session_id=None):
    params = params or {}
    route = route_for(method, routes or DEFAULT_CLIENT_ROUTES)
    if route == "direct_cdp":
        return {"route": route, "target": "direct_cdp", "steps": [{"method": method, "params": params}]}
    if route == "service_worker":
        return {
            "route": route,
            "target": "service_worker",
            "steps": _translate_service_worker_command(method, params, cdp_session_id),
        }
    raise RuntimeError(f"Unsupported client route '{route}' for {method}")


def translate_server_configure(config: dict):
    return {
        "target": "service_worker",
        "steps": [{"method": "Runtime.evaluate", "params": _wrap_magic_configure(config), "unwrap": "evaluate"}],
    }


def unwrap_evaluate_result(result: dict):
    if result.get("exceptionDetails"):
        ex = result["exceptionDetails"]
        msg = (ex.get("exception") or {}).get("description") or ex.get("text") or "Runtime.evaluate failed"
        raise RuntimeError(msg)
    inner = result.get("result") or {}
    return inner.get("value")


def unwrap_binding_called(params: dict, our_session_id: str):
    event = event_name_for(params.get("name") or "")
    if event is None:
        return None
    try:
        payload = json.loads(params.get("payload") or "{}")
    except json.JSONDecodeError:
        return None
    if not isinstance(payload, dict):
        return None
    if our_session_id is not None and payload.get("cdpSessionId") and payload["cdpSessionId"] != our_session_id:
        return None
    data = payload["data"] if "data" in payload else payload
    return {"event": event, "data": data}

File: client/go/translate.go
package modcdp

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

const bindingPrefix = "__ModCDP_"

func DefaultClientRoutes() map[string]string {
	return map[string]string{
		"Mod.*":    "service_worker",
		"Custom.*": "service_worker",
		"*.*":      "service_worker",
	}
}

func bindingNameFor(eventName string) string {
	return bindingPrefix + strings.ReplaceAll(eventName, ".", "_")
}

func eventNameFor(bindingName string) string {
	if !strings.HasPrefix(bindingName, bindingPrefix) {
		return ""
	}
	return strings.ReplaceAll(bindingName[len(bindingPrefix):], "_", ".")
}

func routeFor(method string, routes map[string]string) string {
	if route, ok := routes[method]; ok {
		return route
	}
	bestPrefixLen := -1
	bestRoute := ""
	for pattern, route := range routes {
		if pattern == "*.*" || !strings.HasSuffix(pattern, ".*") {
			continue
		}
		prefix := pattern[:len(pattern)-1]
		if strings.HasPrefix(method, prefix) && len(prefix) > bestPrefixLen {
			bestPrefixLen = len(prefix)
			bestRoute = route
		}
	}
	if bestPrefixLen >= 0 {
		return bestRoute
	}
	if route, ok := routes["*.*"]; ok {
		return route
	}
	return "direct_cdp"
}

type rawStep struct {
	Method string
	Params map[string]any
	Unwrap string
}

type rawCommand struct {
	Route  string
	Target string
	Steps  []rawStep
}

func evalParams(expression string) map[string]any {
	return map[string]any{
		"expression":                  expression,
		"awaitPromise":                true,
		"returnByValue":               true,
		"allowUnsafeEvalBlockedByCSP": true,
	}
}

func wrapModCDPEvaluate(params map[string]any, sessionID string) map[string]any {
	expr, _ := params["expression"].(string)
	userParams := params["params"]
	if userParams == nil {
		userParams = map[string]any{}
	}
	cdpSessionID, _ := params["cdpSessionId"].(string)
	if cdpSessionID == "" {
		cdpSessionID = sessionID
	}
	up, _ := json.Marshal(userParams)
	sid, _ := json.Marshal(cdpSessionID)
	return evalParams(fmt.Sprintf(
		`(async () => { const params = %s; const cdp = globalThis.ModCDP.attachToSession(%s); const ModCDP = globalThis.ModCDP; const chrome = globalThis.chrome; const value = (%s); return typeof value === 'function' ? await value(params) : value; })()`,
		string(up), string(sid), expr,
	))
}

func wrapModCDPAddCustomCommand(params map[string]any) map[string]any {
	name, _ := json.Marshal(params["name"])
	expr, _ := params["expression"].(string)
	exprJSON, _ := json.Marshal(expr)
	pSchema, _ := json.Marshal(params["paramsSchema"])
	rSchema, _ := json.Marshal(params["resultSchema"])
	return evalParams(fmt.Sprintf(
		`(() => { return globalThis.ModCDP.addCustomCommand({ name: %s, paramsSchema: %s, resultSchema: %s, expression: %s, handler: async (params, cdpSessionId, method) => { const cdp = globalThis.ModCDP.attachToSession(cdpSessionId); const ModCDP = globalThis.ModCDP; const chrome = globalThis.chrome; const handler = (%s); return await handler(params || {}, method); }, }); })()`,
		string(name), string(pSchema), string(rSchema), string(exprJSON), expr,
	))
}

func wrapModCDPAddCustomEvent(params map[string]any) map[string]any {
	rawName, _ := params["name"].(string)
	name, _ := json.Marshal(rawName)
	bn, _ := json.Marshal(bindingNameFor(rawName))
	pSchema, _ := json.Marshal(params["eventSchema"])
	return evalParams(fmt.Sprintf(
		`globalThis.ModCDP.addCustomEvent({ name: %s, bindingName: %s, eventSchema: %s })`,
		string(name), string(bn), string(pSchema),
	))
}

func wrapModCDPAddMiddleware(params map[string]any) map[string]any {
	name := params["name"]
	if name == nil {
		name = "*"
	}
	rawExpr, _ := params["expression"].(string)
	nameJSON, _ := json.Marshal(name)
	phaseJSON, _ := json.Marshal(params["phase"])
	exprJSON, _ := json.Marshal(rawExpr)
	return evalParams(fmt.Sprintf(
		`(() => { return globalThis.ModCDP.addMiddleware({ name: %s, phase: %s, expression: %s, handler: async (payload, next, context = {}) => { const cdp = globalThis.ModCDP.attachToSession(context.cdpSessionId ?? null); const ModCDP = globalThis.ModCDP; const chrome = globalThis.chrome; const middleware = (%s); return await middleware(payload, next, context); }, }); })()`,
		string(nameJSON), string(phaseJSON), string(exprJSON), rawExpr,
	))
}

func wrapCustomCommand(method string, params map[string]any, sessionID string) map[string]any {
	m, _ := json.Marshal(method)
	p, _ := json.Marshal(params)
	sid, _ := json.Marshal(sessionID)
	return evalParams(fmt.Sprintf(`globalThis.ModCDP.handleCommand(%s, %s, %s)`, string(m), string(p), string(sid)))
}

func wrapServiceWorkerCommand(method string, params map[string]any, sessionID string) []rawStep {
	if params == nil {
		params = map[string]any{}
	}
	if method == "Mod.ping" {
		if _, ok := params["sentAt"]; !ok {
			next := map[string]any{}
			for key, value := range params {
				next[key] = value
			}
			next["sentAt"] = time.Now().UnixMilli()
			params = next
		}
	}

	if method == "Mod.addCustomEvent" {
		name, _ := params["name"].(string)
		return []rawStep{
			{Method: "Runtime.addBinding", Params: map[string]any{"name": bindingNameFor(name)}},
			{Method: "Runtime.evaluate", Params: wrapModCDPAddCustomEvent(params), Unwrap: "evaluate"},
		}
	}
	runtimeParams := map[string]any{}
	switch method {
	case "Mod.evaluate":
		runtimeParams = wrapModCDPEvaluate(params, sessionID)
	case "Mod.addCustomCommand":
		runtimeParams = wrapModCDPAddCustomCommand(params)
	case "Mod.addMiddleware":
		runtimeParams = wrapModCDPAddMiddleware(params)
	default:
		cdpSessionID, _ := params["cdpSessionId"].(string)
		if cdpSessionID == "" {
			cdpSessionID = sessionID
		}
		runtimeParams = wrapCustomCommand(method, params, cdpSessionID)
	}
	return []rawStep{{Method: "Runtime.evaluate", Params: runtimeParams, Unwrap: "evaluate"}}
}

func wrapCommandIfNeeded(method string, params map[string]any, routes map[string]string, sessionID string) (rawCommand, error) {
	route := routeFor(method, routes)
	if route == "direct_cdp" {
		return rawCommand{Route: route, Target: "direct_cdp", Steps: []rawStep{{Method: method, Params: params}}}, nil
	}
	if route == "service_worker" {
		return rawCommand{Route: route, Target: "service_worker", Steps: wrapServiceWorkerCommand(method, params, sessionID)}, nil
	}
	return rawCommand{}, fmt.Errorf("unsupported client route %q for %s", route, method)
}

func unwrapResponseIfNeeded(result map[string]any, unwrap string) (any, error) {
	if unwrap != "evaluate" {
		return result, nil
	}
	if ex, ok := result["exceptionDetails"].(map[string]any); ok {
		msg := ""
		if e, ok := ex["exception"].(map[string]any); ok {
			if d, ok := e["description"].(string); ok {
				msg = d
			}
		}
		if msg == "" {
			if t, ok := ex["text"].(string); ok {
				msg = t
			}
		}
		if msg == "" {
			msg = "Runtime.evaluate failed"
		}
		return nil, fmt.Errorf("%s", msg)
	}
	inner, _ := result["result"].(map[string]any)
	return inner["value"], nil
}

func unwrapEventIfNeeded(method string, params map[string]any, sessionID string, ourSessionID string) (string, any, bool) {
	if method != "Runtime.bindingCalled" {
		return "", nil, false
	}
	name, _ := params["name"].(string)
	event := eventNameFor(name)
	if event == "" {
		return "", nil, false
	}
	payloadStr, _ := params["payload"].(string)
	var payload map[string]any
	if err := json.Unmarshal([]byte(payloadStr), &payload); err != nil || payload == nil {
		payload = map[string]any{}
	}
	if sid, ok := payload["cdpSessionId"].(string); ok && sid != "" && ourSessionID != "" && sid != ourSessionID {
		return "", nil, false
	}
	if data, ok := payload["data"]; ok {
		return event, data, true
	}
	return event, payload, true
}
